source/material_model/diffusion_dislocation.cc

```diff
@@ -118,27 +135,76 @@ namespace aspect
                             const SymmetricTensor<2,dim> &strain_rate) const
     {
       // Viscosities
-      const double e2inv = second_invariant(strain_rate);
+      // In the first time step, strain rate is normally zero
+      // Sometimes it is also zero at other times.
+      // Let's set it to some very small number
+      double edot;
+      if (strain_rate.norm() == 0.0)
+        {
+          edot = 2.0*std::numeric_limits<double>::min();
+        }
+      else
+        {
+          edot = std::sqrt(abs(second_invariant(strain_rate)));
+        }
+      const double e2inv = edot;
 
       // ---- Find effective viscosities for each of the individual phases
       std::vector<double> composition_viscosities(volume_fractions.size()); // viscosities should have same number of entries as compositional fields
       for (unsigned int j=0; j < volume_fractions.size(); ++j)
         {
-          const double viscosity_diffusion = std::min(1e22,(1e0/prefactors_diffusion[j])*
-                                                      std::exp((activation_energies_diffusion[j]+activation_volumes_diffusion[j]*pressure)
-                                                               /(constants::gas_constant*temperature)));
-
-          double one_over_viscosity_dislocation = 0.0;
-          if (e2inv > 2.0*std::numeric_limits<double>::min())
-            one_over_viscosity_dislocation = (constants::gas_constant*temperature)
-                                             /
-                                             std::min(1e22,std::pow(prefactors_dislocation[j],-1e0/stress_exponents_dislocation[j])*
-                                                      std::pow(e2inv,(1e0-stress_exponents_dislocation[j])/
-                                                               stress_exponents_dislocation[j])*
-                                                      std::exp((activation_energies_dislocation[j]+
-                                                                activation_volumes_dislocation[j]*pressure)/(stress_exponents_dislocation[j])));
-
-          composition_viscosities[j] = std::min(std::max(std::pow((1.0/viscosity_diffusion + one_over_viscosity_dislocation), -1.0), min_visc), max_visc);
+          // Power law creep equation
+          // \eta = 0.5 * A_i^{-\frac{1}{n_i}} d^\frac{m}{n} e2inv_i^{\frac{1-n_i}{n_i}} \exp\left(\frac{E_i^* + PV_i^*}{n_iRT}\right)
+          // where i corresponds to diffusion or dislocation creep
+
+          double e2inv_diffusion = 0;
+          double e2inv_dislocation = e2inv;
```

Comment:
why do you need these two variables? I don't see you modifying them?

These and several of the variables below can be made `const` as far as I can see.